flecsi/exec/hpx/policy.hh

```diff
@@ -0,0 +1,193 @@
+reduce_internal(Args &&... args) {
+  using Traits = util::function_traits<decltype(F)>;
+  using R = typename Traits::return_type;
+
+  // replace arguments in args, for example, field_reference -> accessor.
+  auto params = exec::detail::replace_arguments(
+    static_cast<typename Traits::arguments_type *>(nullptr),
+    std::forward<Args>(args)...);
+
+  auto task_name = util::symbol<F>();
+
+  // Now we have accessors, we need to bind the accessor to real memory
+  // for the data field. We also need to patch up default conversion
+  // from args to params, especially for the future<>.
+  // This is being achieved by creating the prolog<> instances below.
+
+  // Different kinds of task invocation with flecsi::execute():
+  // 1. domain_size is a std::monostate: a single task launch. On a single
+  //    rank (0) of our choice, we apply F to ARGS. Given the return type R of
+  //    F, we return a future<R, single>. The client can later call
+  //    future<>.get() on any rank to get the result. The value returned by
+  //    .get should be the same on all ranks, implying a Broadcast is needed.
+  // 2. Reduction is not void: a true reduction task.
+  //    We apply F to ARGS and reduce the return values with the Reduction
+  //    operation. We then put the reduced value into a future<R,
+  //    launch_type::single> and return it. Again, the client could call
+  //    .get() to get the (same) result on any rank, implying either an
+  //    Allreduce or Reduce/Broadcast is needed.
+  // 3. Reduction is void: an index launch. We apply F to ARGS on every rank,
+  //    each will return a value r_i. The value r_i will be put into the the
+  //    slot i in the future<R, index>. The client can call future<R,
+  //    index>.get(j) to get the return value on any rank j. This implies an
+  //    Allgather is needed.
+  const auto ds = launch_size<Attributes, decltype(params)>(args...);
+  util::annotation::rguard<util::annotation::execute_task_user> ann{task_name};
+
+  // The prolog will calculate dependencies between tasks based on the
+  // attributes associated with the arguments.
+  prolog<mask_to_processor_type(Attributes)> bound_params(params, args...);
+
+  const auto delay = [&](auto && f) {
+    return std::move(bound_params)
+      .template delay_execution<R>(std::move(params),
+        std::move(task_name),
+        std::forward<decltype(f)>(f),
+        std::move(args)...);
+  };
+
+  using delayed_prolog =
+    prolog<mask_to_processor_type(Attributes), exec::delayed_task_prologue>;
+
+  if constexpr(std::is_same_v<decltype(ds), const std::monostate>) {
+    const bool root = flecsi::run::context::instance().process() == 0;
+
+    // single launch, only invoke the user task on the Root.
+    if constexpr(std::is_void_v<R>) {
+      if(root) {
+        return future<void>{delay([](auto && params, auto &&... args) {
+          // void return type, just invoke, no return value to broadcast
+          delayed_prolog(params, args...);
```

Comment:
Whether this ends up using `args...` or not, it could probably be factored out into `delay` via yet another wrapper lambda.